### ISSUE TITLE
Memory leak calculating SHA256 hash of large file

### ISSUE THREAD

**Issue by sunrei** [vibornoff/asmcrypto.js#69]
I'm trying to develope an app using https://github.com/MediaFire/mediafire-javascript-sdk
The SDK is used for uploading a file to MediaFire server. But before uploading it calculates SHA256 hash with asmCrypto.js in file https://github.com/MediaFire/mediafire-javascript-sdk/blob/master/sha256.js
Generally I have a file with size of 4,6GB.
It is being splitted into 16MB sections. For each section we calculate SHA256 hash (asmCrypto.SHA256.hex on ArrayBuffer) and calculate total file hash as progressive operation (asmCrypto.SHA256.init , then 'process' for each section and finish().result at the end).
And these operations result in memory leak leading to 'out of memory' failure in the middle of the process on Windows.
I have tested this on 6 workstations:
1) Laptop with Windows XP SP3, 32bit, 2GB RAM - Fails
2) PC with Windows 7, 32bit, 3 GB RAM - Fails
3) PC with Windows 7, 64bit, 4 GB RAM - Fails
4) PC with ubuntu 14.04. 64bit, 6GB RAM - OK
5) Another PC with ubuntu 14.04. 64bit, 6GB RAM - OK
6) Macbook Pro, 8GB RAM - OK

The screenshot is taken from 3rd configuration 
![crash](https://cloud.githubusercontent.com/assets/653959/6214201/6e52086e-b5ff-11e4-8c6b-fc6d1e1ff007.jpg)

I have also tried latest version of asmcrypto.js , but even without progressive operation (just calculating hashes for each unit) it fails but at later point.

**Comment by alippai** [comment]
Did you use Chrome? Can you try this in Firefox as well?

**Comment by vibornoff** [comment]
Could you please prepare a sandbox like [this one](https://grafilltest.alreadyon.com/mediatest/upload3.php) with updated development build of asmcrypto.js ?

**Comment by sunrei** [comment]
alippai, I've tried both Chrome and FF. The screenshot above is from chrome. FF actually does not show errors, but it stops receiving messages from web worker near the same point.
I've also tried in IE 11, and it throws error in console like "out of memory".

vibornoff, I've just prepared a new sandbox at following url https://grafilltest.alreadyon.com/mediatest/upload4.php
But I had to disable code for progressive SHA256 calculation of the file. So it just calculating hashes of each chunk separately. 

I've also just tried running tests on laptop with lubuntu (32bit, 2GB RAM minus 128MB video memory).
![2015-02-16-223653_1280x800_scrot](https://cloud.githubusercontent.com/assets/653959/6219090/1028b2ea-b62e-11e4-9a00-2d8dccf46c01.png)

Results are almost the same as on Windows. Script fails after about 200-300 Mb of processed data.
Total memory usage of webworker with hash calculating script is above 1,5 GB that I consider enormous.

I must admit that CryptoJS, that is used for example on https://md5file.com/calculator
does not have this issue. It can process large files, but the speed is like 10 times worse than in asmCrypto.js.

I understand that approaches are completely different. And I personally would like to help with fixing asmCrypto in any possible ways I can. But unfortunately I'm not good with low level engineering :(

**Comment by vibornoff** [comment]
Seems you overfeeding the worker.
I took hasher.js and sha256.js unchanged from your setup and run along with the different index page:
```html
<html>
    <head>
        <title>memleak test</title>
    </head>
    <body>
        <button id="b">Start</button>
        <script>
            var b = document.getElementById('b'), w;
            b.onclick = function () {
                if ( w ) {
                    w.terminate(), w = undefined;
                    b.innerText = "Start";
                }
                else {
                    w = new Worker("hasher.js");
                    w.u = 1000;
                    w.postMessage(w.u);
                    w.onmessage = function ( e ) {
                        if ( --w.u ) {
                            w.next();
                        } else {
                            b.onclick();
                        }
                        console.log( w.u, JSON.stringify(e.data) );
                    };
                    w.next = function () {
                        var a = new ArrayBuffer(16*1024*1024);
                        w.postMessage(a,[a]);
                    };
                    w.next();
                    b.innerText = "Stop";
                }
            }
        </script>
    </body>
</html>
```
My setup doesn't seem draining memory.

P.S. Use newer version of asmcrypto.js. Your sha256.js is... I don't even remember the moment when it was looking like you have in your setup.

**Comment by sunrei** [comment]
Thank you for your response. Seems like I found the reason of an issue.
It's not actually in hash script itself.

Original script was posting messages to web worker as soon as FileReader read each chunk of data from file without waiting for response from web worker. 
Web worker has to pile up received data because file reading was faster than hash calculation. Without instant garbage collection that looked like memory leak.
Finally when the system was out of memory because of storing queued data web worker was crashing.

I have mistakenly decided that this issue was related to hash algorithm because when I turned out hash calculations web worker did not drain much memory. But that was explained in instant back messages from web worker and system did not have to queue data.

So my solution was to send next chunk of data to web worker only after successful processing of previous chunk. 

Thank you very much! I really appreciate your help.
